File: bindings/typescript/src/verify.ts
import type { AuthsIssue, Receipt } from "./index.js";
import { issue } from "./internal/issues.js";
import { parsePortableReceipt } from "./internal/receipt.js";
import { loadPackagedWorkflowEngine, type PackagedWorkflowEngine } from "./verifier/wasm.js";
import {
  loadVerifier,
  type VerificationResult as NativeVerificationResult,
  type Verifier as NativeVerifier,
} from "./verifier/client.js";

export type VerificationStage = "decode" | "resolve" | "principal-control" | "authority" | "complete";

export interface VerificationInput {
  readonly proof: Uint8Array;
  readonly action: Uint8Array;
  readonly trustedContext: Uint8Array;
}

export interface VerificationMetrics {
  readonly proofBytes: bigint;
  readonly actionBytes: bigint;
  readonly contextBytes: bigint;
  readonly objectCount: bigint;
  readonly planLeaves: bigint;
  readonly planDepth: bigint;
  readonly workUnits: bigint;
}

interface VerificationCommon {
  readonly code: string;
  readonly stage: VerificationStage;
  readonly correlationId: string;
  readonly metrics: VerificationMetrics;
  readonly requiredConfiguration?: Uint8Array;
  readonly executedConfiguration: Uint8Array;
  readonly decisionBytes: Uint8Array;
}

export type VerificationResult =
  | (VerificationCommon & Readonly<{ kind: "authorized" }>)
  | (VerificationCommon & Readonly<{ kind: "denied"; issue: AuthsIssue & Readonly<{ effect: "not-applied" }> }>)
  | (VerificationCommon & Readonly<{ kind: "indeterminate"; issue: AuthsIssue & Readonly<{ effect: "not-applied" }> }>);

export interface VerificationInspection {
  readonly kind: VerificationResult["kind"];
  readonly code: string;
  readonly stage: VerificationStage;
  readonly resultCommitment: Uint8Array;
  readonly actionCommitment?: Uint8Array;
  readonly requiredConfigurationCommitment?: Uint8Array;
  readonly executedConfigurationCommitment: Uint8Array;
  readonly metrics: VerificationMetrics;
  readonly approval?: Readonly<{
    policyId: string;
    evaluatorVersion: string;
    decision: "approved" | "rejected";
    commitment: Uint8Array;
  }>;
}

export interface ReceiptProfile { readonly id: string; readonly version: number }
export interface ReceiptTrustAnchor {
  readonly role: "decision" | "execution";
  readonly principal: string;
  readonly verificationMethod: string;
  readonly suite: "ed25519-v1" | "p256-sha256-v1";
  readonly publicKey: Uint8Array;
}

declare const receiptTrustPolicyBrand: unique symbol;
export interface ReceiptTrustPolicy {
  readonly [receiptTrustPolicyBrand]: true;
  readonly allowedProfiles: readonly ReceiptProfile[];
  readonly anchorCount: number;
}

interface InternalReceiptTrustPolicy extends ReceiptTrustPolicy {
  readonly anchors: readonly ReceiptTrustAnchor[];
  readonly verificationTimeUnixSeconds?: bigint;
  readonly maximumReceiptAgeSeconds: bigint;
  readonly engine: PackagedWorkflowEngine;
}

const trustPolicies = new WeakSet<object>();

export async function pinnedReceiptTrust(options: Readonly<{
  anchors: readonly ReceiptTrustAnchor[];
  allowedProfiles: readonly ReceiptProfile[];
  verificationTimeUnixSeconds?: bigint;
  maximumReceiptAgeSeconds?: bigint;
}>): Promise<ReceiptTrustPolicy> {
  if (options.anchors.length < 1 || options.anchors.length > 32 ||
      !options.anchors.some((anchor) => anchor.role === "decision")) {
    throw new RangeError("receipt trust requires 1..32 anchors including a decision anchor");
  }
  if (options.allowedProfiles.length < 1 || options.allowedProfiles.length > 16) {
    throw new RangeError("receipt trust requires 1..16 profiles");
  }
  const maximumReceiptAgeSeconds = options.maximumReceiptAgeSeconds ?? 86_400n;
  if (maximumReceiptAgeSeconds < 1n || maximumReceiptAgeSeconds > 31_536_000n) {
    throw new RangeError("maximum receipt age is outside bounds");
  }
  if (options.verificationTimeUnixSeconds !== undefined && options.verificationTimeUnixSeconds < 0n) {
    throw new RangeError("verification time is invalid");
  }
  const engine = await loadPackagedWorkflowEngine();
  const profiles = Object.freeze(options.allowedProfiles.map(parseProfile));
  if (new Set(profiles.map((profile) => `${profile.id}/${profile.version}`)).size !== profiles.length) {
    throw new TypeError("duplicate receipt profile");
  }
  const anchors = Object.freeze(options.anchors.map((anchor) => {
    const parsed = parseAnchor(anchor);
    engine.validateReceiptAnchorV1(parsed.suite, parsed.publicKey);
    return parsed;
  }));
  if (new Set(anchors.map((anchor) => `${anchor.role}\0${anchor.principal}\0${anchor.verificationMethod}\0${anchor.suite}`)).size !== anchors.length) {
    throw new TypeError("duplicate receipt trust anchor");
  }
  const value = Object.freeze({
    allowedProfiles: profiles,
    anchorCount: anchors.length,
    anchors,
    engine,
    maximumReceiptAgeSeconds,
    ...(options.verificationTimeUnixSeconds === undefined
      ? {}
      : { verificationTimeUnixSeconds: options.verificationTimeUnixSeconds }),
  }) as unknown as InternalReceiptTrustPolicy;
  trustPolicies.add(value);
  return value;
}

export interface DecisionReceiptDetails {
  readonly kind: "decision";
  readonly receiptId: string;
  readonly profile: ReceiptProfile;
  readonly decision: "authorized" | "denied" | "indeterminate";
  readonly reasons: readonly string[];
  readonly decidedAtUnixSeconds: bigint;
  readonly decisionSigner: Readonly<{ principal: string; verificationMethod: string; suite: string }>;
  readonly commitments: Readonly<{ proof: string; action: string; context: string; principalStatus: string; grantStatus: string }>;
}

export interface ExecutionReceiptDetails {
  readonly kind: "execution";
  readonly decisionReceiptId: string;
  readonly executionReceiptId: string;
  readonly profile: ReceiptProfile;
  readonly decision: "authorized" | "denied" | "indeterminate";
  readonly outcome: "succeeded" | "failed" | "indeterminate";
  readonly reasons: readonly string[];
  readonly decidedAtUnixSeconds: bigint;
  readonly completedAtUnixSeconds: bigint;
  readonly decisionSigner: Readonly<{ principal: string; verificationMethod: string; suite: string }>;
  readonly executionSigner: Readonly<{ principal: string; verificationMethod: string; suite: string }>;
  readonly commitments: Readonly<{ proof: string; action: string; context: string; principalStatus: string; grantStatus: string; executionLease: string; command: string; result?: string }>;
}

export type ReceiptEnvelopeDetails = DecisionReceiptDetails | ExecutionReceiptDetails;
declare const verifiedReceiptBrand: unique symbol;
export interface VerifiedReceipt {
  readonly [verifiedReceiptBrand]: true;
  readonly kind: "verified";
  readonly receipt: Receipt;
  readonly details: ReceiptEnvelopeDetails;
}

export type ReceiptVerification =
  | VerifiedReceipt
  | Readonly<{ kind: "rejected"; issue: AuthsIssue & Readonly<{ effect: "not-applied" }> }>
  | Readonly<{ kind: "indeterminate"; issue: AuthsIssue & Readonly<{ effect: "not-applied" }> }>;

export interface VerificationOptions { readonly correlationId?: string }
export interface VerificationBatchOptions {
  readonly signal?: AbortSignal;
  readonly chunkSize?: number;
  readonly correlationId?: () => string;
}
export interface ReceiptVerificationInput {
  readonly receipt: Receipt | Uint8Array;
  readonly trust: ReceiptTrustPolicy;
}

export class Verifier {
  readonly #native: NativeVerifier;
  private constructor(native: NativeVerifier) { this.#native = native; }

  static async create(): Promise<Verifier> { return new Verifier(await loadVerifier()); }

  verify(input: VerificationInput, options: VerificationOptions = {}): VerificationResult {
    validateInput(input);
    return projectResult(this.#native.verify(input.proof, input.action, input.trustedContext, options));
  }

  async verifyMany(inputs: readonly VerificationInput[], options: VerificationBatchOptions = {}): Promise<readonly VerificationResult[]> {
    inputs.forEach(validateInput);
    const results = await this.#native.verifyMany(inputs.map((input) => ({
      proofCbor: input.proof,
      canonicalActionCbor: input.action,
      trustedContextCbor: input.trustedContext,
    })), options);
    return Object.freeze(results.map(projectResult));
  }

  inspect(result: VerificationResult): VerificationInspection {
    return Object.freeze({
      kind: result.kind,
      code: result.code,
      stage: result.stage,
      resultCommitment: commitment("result", result.decisionBytes),
      ...(result.kind === "authorized" ? { actionCommitment: commitment("action", result.decisionBytes) } : {}),
      ...(result.requiredConfiguration === undefined ? {} : { requiredConfigurationCommitment: commitment("required", result.requiredConfiguration) }),
      executedConfigurationCommitment: commitment("executed", result.executedConfiguration),
      metrics: result.metrics,
    });
  }

  verifyReceipt(input: ReceiptVerificationInput): ReceiptVerification {
    if (!trustPolicies.has(input.trust as object)) throw new TypeError("unsealed receipt trust policy");
    try {
      const trust = input.trust as InternalReceiptTrustPolicy;
      const receipt = parsePortableReceipt(input.receipt, trust.engine);
      const decision = verifyWithAnchors(
        trust,
        "decision",
        receipt.attestedDecision,
        receipt.decisionReceiptId,
      );
      let metadata = decision;
      if (receipt.kind === "execution") {
        if (receipt.attestedExecution === undefined || receipt.executionReceiptId === undefined) {
          throw new TypeError("execution receipt omitted its execution attestation");
        }
        metadata = verifyWithAnchors(
          trust,
          "execution",
          receipt.attestedExecution,
          receipt.executionReceiptId,
        );
        trust.engine.verifyReceiptLinkV1(
          receipt.attestedDecision,
          receipt.decisionReceiptId,
          receipt.attestedExecution,
          receipt.executionReceiptId,
        );
        if (metadata.decisionReceiptId !== hex(receipt.decisionReceiptId)) throw new TypeError("receipt chain mismatch");
      }
      const profile = parseProfile(decision.profile as ReceiptProfile);
      if (!trust.allowedProfiles.some((allowed) => allowed.id === profile.id && allowed.version === profile.version)) {
        return Object.freeze({ kind: "rejected", issue: notAppliedIssue("core.receipt-profile-denied") });
      }
      const now = trust.verificationTimeUnixSeconds ?? BigInt(Math.floor(Date.now() / 1000));
      const decidedAt = BigInt(String(decision.decidedAtUnixSeconds));
      const completedAt = receipt.kind === "execution" ? BigInt(String(metadata.completedAtUnixSeconds)) : decidedAt;
      if (decidedAt > now + 300n || completedAt > now + 300n || now - decidedAt > trust.maximumReceiptAgeSeconds) {
        return Object.freeze({ kind: "rejected", issue: notAppliedIssue("core.receipt-expired") });
      }
      const common = decision.commitments as Record<string, string>;
      const detailsValue: Record<string, unknown> = receipt.kind === "decision"
        ? {
            kind: "decision", receiptId: hex(receipt.decisionReceiptId), profile,
            decision: decision.decision, reasons: Object.freeze([...decision.reasons]),
            decidedAtUnixSeconds: decidedAt, decisionSigner: Object.freeze({ ...decision.decisionSigner }),
            commitments: Object.freeze({ proof: common.proof!, action: common.action!, context: common.context!, principalStatus: common.principalStatus!, grantStatus: common.grantStatus! }),
          }
        : {
            kind: "execution", decisionReceiptId: String(metadata.decisionReceiptId), executionReceiptId: hex(receipt.executionReceiptId!), profile,
            decision: decision.decision, outcome: metadata.outcome, reasons: Object.freeze([...decision.reasons]),
            decidedAtUnixSeconds: decidedAt, completedAtUnixSeconds: completedAt,
            decisionSigner: Object.freeze({ ...decision.decisionSigner }), executionSigner: Object.freeze({ ...metadata.executionSigner }),
            commitments: Object.freeze({ proof: common.proof!, action: common.action!, context: common.context!, principalStatus: common.principalStatus!, grantStatus: common.grantStatus!, executionLease: metadata.commitments.executionLease, command: metadata.commitments.command, ...(metadata.commitments.result === null ? {} : { result: metadata.commitments.result }) }),
          };
      const details = Object.freeze(detailsValue) as unknown as ReceiptEnvelopeDetails;
      return Object.freeze({ kind: "verified", receipt: receipt.receipt, details }) as VerifiedReceipt;
    } catch {
      return Object.freeze({ kind: "rejected", issue: notAppliedIssue("core.receipt-malformed") });
    }
  }
}

export async function createVerifier(): Promise<Verifier> { return Verifier.create(); }

function projectResult(result: NativeVerificationResult): VerificationResult {
  const common = {
    code: result.code,
    stage: result.stage,
    correlationId: result.correlationId,
    metrics: result.metrics,
    ...(result.requiredConfiguration === undefined ? {} : { requiredConfiguration: result.requiredConfiguration.slice() }),
    executedConfiguration: result.localConfiguration.slice(),
    decisionBytes: result.resultCbor.slice(),
  };
  if (result.kind === "authorized") return Object.freeze({ ...common, kind: "authorized" });
  const projectedIssue = notAppliedIssue(
    result.kind === "denied" ? "core.authorization-denied" : "core.authorization-indeterminate",
    { correlationId: result.correlationId, summary: result.explanation.message },
  );
  return result.kind === "denied"
    ? Object.freeze({ ...common, kind: "denied", issue: projectedIssue })
    : Object.freeze({ ...common, kind: "indeterminate", issue: projectedIssue });
}

function notAppliedIssue(
  code: Parameters<typeof issue>[0],
  options?: Parameters<typeof issue>[1],
): AuthsIssue & Readonly<{ effect: "not-applied" }> {
  const value = issue(code, options);
  if (value.effect !== "not-applied") throw new TypeError("expected a not-applied registry code");
  return value as AuthsIssue & Readonly<{ effect: "not-applied" }>;
}

function validateInput(input: VerificationInput): void {
  if (!(input.proof instanceof Uint8Array) || !(input.action instanceof Uint8Array) || !(input.trustedContext instanceof Uint8Array)) throw new TypeError("verification input must contain bytes");
}

function parseProfile(value: ReceiptProfile): ReceiptProfile {
  if (!/^[a-z][a-z0-9._:/-]{0,127}$/u.test(value.id) || !Number.isSafeInteger(value.version) || value.version < 1 || value.version > 0x7fffffff) throw new TypeError("invalid receipt profile");
  return Object.freeze({ id: value.id, version: value.version });
}

function parseAnchor(anchor: ReceiptTrustAnchor): ReceiptTrustAnchor {
  if (!/^[a-z][a-z0-9._:/-]{0,127}$/u.test(anchor.suite) || anchor.principal.length < 1 || anchor.principal.length > 512 || anchor.verificationMethod.length < 1 || anchor.verificationMethod.length > 512) throw new TypeError("invalid receipt anchor");
  const expected = anchor.suite === "ed25519-v1" ? 32 : 33;
  if (!(anchor.publicKey instanceof Uint8Array) || anchor.publicKey.length !== expected) throw new TypeError("invalid receipt public key");
  return Object.freeze({ ...anchor, publicKey: anchor.publicKey.slice() });
}

function commitment(domain: string, bytes: Uint8Array): Uint8Array {
  const output = new Uint8Array(32);
  let state = 0x811c9dc5;
  for (const byte of new TextEncoder().encode(domain)) { state ^= byte; state = Math.imul(state, 0x01000193); }
  for (const byte of bytes) { state ^= byte; state = Math.imul(state, 0x01000193); }
  for (let index = 0; index < output.length; index += 1) { state ^= index; state = Math.imul(state, 0x01000193); output[index] = state >>> ((index % 4) * 8); }
  return output;
}

function verifyWithAnchors(
  trust: InternalReceiptTrustPolicy,
  kind: "decision" | "execution",
  attested: Uint8Array,
  receiptId: Uint8Array,
): any {
  const role = kind;
  for (const anchor of trust.anchors) {
    if (anchor.role !== role) continue;
    try {
      const encoded = trust.engine.verifyPinnedReceiptV1(kind, attested, receiptId, anchor.principal, anchor.verificationMethod, anchor.suite, anchor.publicKey);
      return JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(encoded));
    } catch {
      continue;
    }
  }
  throw new TypeError("receipt signer is not pinned or its signature is invalid");
}

function hex(value: Uint8Array): string {
  return Array.from(value, (byte) => byte.toString(16).padStart(2, "0")).join("");
}

File: bindings/python/python/auths/verify.py
from __future__ import annotations

import hashlib
import json
import time
from dataclasses import dataclass
from typing import Any, Dict, Iterable, Literal, Optional, Tuple, Union

from . import ErrorInfo, Receipt
from ._native import (
    NativeVerificationResult, validate_receipt_anchor_v1,
    verify_many_v1, verify_pinned_receipt_v1, verify_receipt_link_v1, verify_v1,
)
from ._public import error_info, mint_receipt, parse_portable_receipt

VerificationStage = Literal["decode", "resolve", "principal-control", "authority", "complete"]
VerificationKind = Literal["authorized", "denied", "indeterminate"]


@dataclass(frozen=True, init=False)
class VerificationInput:
    proof: bytes
    action: bytes
    trusted_context: bytes

    def __init__(self, *, proof: bytes, action: bytes, trusted_context: bytes) -> None:
        object.__setattr__(self, "proof", bytes(proof))
        object.__setattr__(self, "action", bytes(action))
        object.__setattr__(self, "trusted_context", bytes(trusted_context))


@dataclass(frozen=True)
class VerificationMetrics:
    proof_bytes: int
    action_bytes: int
    context_bytes: int
    object_count: int
    plan_leaves: int
    plan_depth: int
    work_units: int


@dataclass(frozen=True)
class AuthorizedVerification:
    kind: Literal["authorized"]
    code: str
    stage: VerificationStage
    correlation_id: str
    metrics: VerificationMetrics
    required_configuration: Optional[bytes]
    executed_configuration: bytes
    decision_bytes: bytes


@dataclass(frozen=True)
class UnsuccessfulVerification:
    kind: Literal["denied", "indeterminate"]
    code: str
    stage: VerificationStage
    correlation_id: str
    metrics: VerificationMetrics
    required_configuration: Optional[bytes]
    executed_configuration: bytes
    decision_bytes: bytes
    issue: ErrorInfo


VerificationResult = Union[AuthorizedVerification, UnsuccessfulVerification]


@dataclass(frozen=True)
class ApprovalInspection:
    policy_id: str
    evaluator_version: str
    decision: Literal["approved", "rejected"]
    commitment: bytes


@dataclass(frozen=True)
class VerificationInspection:
    kind: VerificationKind
    code: str
    stage: VerificationStage
    result_commitment: bytes
    action_commitment: Optional[bytes]
    required_configuration_commitment: Optional[bytes]
    executed_configuration_commitment: bytes
    metrics: VerificationMetrics
    approval: Optional[ApprovalInspection]


@dataclass(frozen=True)
class ReceiptSignerInfo:
    principal: str
    verification_method: str
    suite: str


ReceiptSignerRole = Literal["decision", "execution"]


@dataclass(frozen=True)
class ReceiptProfile:
    id: str
    version: int


@dataclass(frozen=True)
class ReceiptTrustAnchor:
    role: ReceiptSignerRole
    principal: str
    verification_method: str
    suite: Literal["ed25519-v1", "p256-sha256-v1"]
    public_key: bytes


_TRUST_TOKEN = object()


class ReceiptTrustPolicy:
    def __init__(self, token: object, anchors: Tuple[ReceiptTrustAnchor, ...], profiles: Tuple[ReceiptProfile, ...], verification_time: Optional[int], maximum_age: int) -> None:
        if token is not _TRUST_TOKEN:
            raise TypeError("ReceiptTrustPolicy is sealed")
        self._anchors = anchors
        self._profiles = profiles
        self._verification_time = verification_time
        self._maximum_age = maximum_age

    @property
    def allowed_profiles(self) -> Tuple[ReceiptProfile, ...]:
        return self._profiles

    @property
    def anchor_count(self) -> int:
        return len(self._anchors)


def pinned_receipt_trust(*, anchors: Iterable[ReceiptTrustAnchor], allowed_profiles: Iterable[ReceiptProfile], verification_time_unix_seconds: Optional[int] = None, maximum_receipt_age_seconds: Optional[int] = None) -> ReceiptTrustPolicy:
    anchor_values = tuple(ReceiptTrustAnchor(value.role, value.principal, value.verification_method, value.suite, bytes(value.public_key)) for value in anchors)
    profiles = tuple(ReceiptProfile(value.id, value.version) for value in allowed_profiles)
    if not 1 <= len(anchor_values) <= 32 or not any(value.role == "decision" for value in anchor_values):
        raise ValueError("receipt trust requires 1..32 anchors including a decision anchor")
    if not 1 <= len(profiles) <= 16:
        raise ValueError("receipt trust requires 1..16 profiles")
    if len({(value.id, value.version) for value in profiles}) != len(profiles):
        raise ValueError("duplicate receipt profile")
    if verification_time_unix_seconds is not None and verification_time_unix_seconds < 0:
        raise ValueError("verification time is invalid")
    seen = set()
    for anchor in anchor_values:
        identity = (anchor.role, anchor.principal, anchor.verification_method, anchor.suite)
        if identity in seen:
            raise ValueError("duplicate receipt trust anchor")
        seen.add(identity)
        validate_receipt_anchor_v1(anchor.suite, anchor.public_key)
    maximum_age = 86_400 if maximum_receipt_age_seconds is None else maximum_receipt_age_seconds
    if not 1 <= maximum_age <= 31_536_000:
        raise ValueError("maximum receipt age is outside bounds")
    return ReceiptTrustPolicy(_TRUST_TOKEN, anchor_values, profiles, verification_time_unix_seconds, maximum_age)


@dataclass(frozen=True)
class DecisionReceiptDetails:
    kind: Literal["decision"]
    receipt_id: str
    profile_id: str
    profile_version: int
    decision: Literal["authorized", "denied", "indeterminate"]
    reasons: Tuple[str, ...]
    decided_at_unix_seconds: int
    decision_signer: ReceiptSignerInfo
    proof_commitment: str
    action_commitment: str
    context_commitment: str
    principal_status_commitment: str
    grant_status_commitment: str


@dataclass(frozen=True)
class ExecutionReceiptDetails:
    kind: Literal["execution"]
    decision_receipt_id: str
    execution_receipt_id: str
    profile_id: str
    profile_version: int
    decision: Literal["authorized", "denied", "indeterminate"]
    outcome: Literal["succeeded", "failed", "indeterminate"]
    reasons: Tuple[str, ...]
    decided_at_unix_seconds: int
    completed_at_unix_seconds: int
    decision_signer: ReceiptSignerInfo
    execution_signer: ReceiptSignerInfo
    proof_commitment: str
    action_commitment: str
    context_commitment: str
    principal_status_commitment: str
    grant_status_commitment: str
    execution_lease_commitment: str
    command_commitment: str
    result_commitment: Optional[str]


ReceiptEnvelopeDetails = Union[DecisionReceiptDetails, ExecutionReceiptDetails]


@dataclass(frozen=True, init=False)
class VerifiedReceipt:
    kind: Literal["verified"]
    receipt: Receipt
    details: ReceiptEnvelopeDetails

    def __init__(self, token: object, receipt: Receipt, details: ReceiptEnvelopeDetails) -> None:
        if token is not _VERIFIED_TOKEN:
            raise TypeError("VerifiedReceipt is sealed")
        object.__setattr__(self, "kind", "verified")
        object.__setattr__(self, "receipt", receipt)
        object.__setattr__(self, "details", details)


@dataclass(frozen=True)
class RejectedReceipt:
    kind: Literal["rejected"]
    issue: ErrorInfo


@dataclass(frozen=True)
class IndeterminateReceipt:
    kind: Literal["indeterminate"]
    issue: ErrorInfo


ReceiptVerification = Union[VerifiedReceipt, RejectedReceipt, IndeterminateReceipt]
_VERIFIED_TOKEN = object()


class _Verifier:
    def __init__(self, token: object) -> None:
        if token is not _VERIFIER_TOKEN:
            raise TypeError("Verifier is sealed")

    def verify(self, input: VerificationInput, *, correlation_id: Optional[str] = None) -> VerificationResult:
        native = verify_v1(input.proof, input.action, input.trusted_context)
        return _project(native, correlation_id or f"auths-{time.time_ns():x}")

    def verify_many(self, inputs: Iterable[VerificationInput], *, chunk_size: int = 32) -> Tuple[VerificationResult, ...]:
        values = tuple(inputs)
        if not 1 <= len(values) <= 256 or not 1 <= chunk_size <= 256:
            raise ValueError("verification batch is outside bounds")
        native = verify_many_v1([(value.proof, value.action, value.trusted_context) for value in values])
        return tuple(_project(value, f"auths-{time.time_ns():x}-{index}") for index, value in enumerate(native))

    def inspect(self, result: VerificationResult) -> VerificationInspection:
        def digest(domain: bytes, value: bytes) -> bytes:
            return hashlib.sha256(domain + value).digest()
        return VerificationInspection(result.kind, result.code, result.stage, digest(b"result", result.decision_bytes), digest(b"action", result.decision_bytes) if result.kind == "authorized" else None, None if result.required_configuration is None else digest(b"required", result.required_configuration), digest(b"executed", result.executed_configuration), result.metrics, None)

    def verify_receipt(self, *, receipt: Union[Receipt, bytes], trust: ReceiptTrustPolicy) -> ReceiptVerification:
        if not isinstance(trust, ReceiptTrustPolicy):
            raise TypeError("unsealed receipt trust policy")
        try:
            if not isinstance(receipt, (Receipt, bytes)):
                raise TypeError("receipt must be sealed or bytes")
            source_bytes = receipt.to_bytes() if isinstance(receipt, Receipt) else receipt
            kind, portable_id, decision_id, execution_id, attested_decision, attested_execution = parse_portable_receipt(source_bytes)
            source = receipt if isinstance(receipt, Receipt) else mint_receipt(portable_id, source_bytes)
            if source.id != portable_id:
                raise ValueError
            decision_metadata = _verify_with_anchors(
                "decision", attested_decision, decision_id, "decision", trust,
            )
            metadata = decision_metadata
            if kind == "execution":
                if execution_id is None or attested_execution is None:
                    raise ValueError
                metadata = _verify_with_anchors(
                    "execution", attested_execution, execution_id, "execution", trust,
                )
                verify_receipt_link_v1(
                    attested_decision, decision_id, attested_execution, execution_id,
                )
                if metadata["decisionReceiptId"] != decision_id.hex():
                    raise ValueError

            profile = decision_metadata["profile"]
            if not any(value.id == profile["id"] and value.version == profile["version"] for value in trust.allowed_profiles):
                return RejectedReceipt("rejected", error_info("core.receipt-profile-denied"))
            now = trust._verification_time
            if now is None:
                try: now = int(time.time())
                except (OSError, OverflowError):
                    return IndeterminateReceipt("indeterminate", error_info("core.receipt-trust-indeterminate"))
            decided_at = int(decision_metadata["decidedAtUnixSeconds"])
            completed_at = decided_at if kind == "decision" else int(metadata["completedAtUnixSeconds"])
            if decided_at > now + 300 or completed_at > now + 300 or now - decided_at > trust._maximum_age:
                return RejectedReceipt("rejected", error_info("core.receipt-expired"))
            decision_signer = _signer(decision_metadata["decisionSigner"])
            common = decision_metadata["commitments"]
            if kind == "decision":
                details: ReceiptEnvelopeDetails = DecisionReceiptDetails(
                    "decision", decision_id.hex(), profile["id"], int(profile["version"]),
                    decision_metadata["decision"], tuple(decision_metadata["reasons"]), decided_at,
                    decision_signer, common["proof"], common["action"], common["context"],
                    common["principalStatus"], common["grantStatus"],
                )
            else:
                assert execution_id is not None
                execution_commitments = metadata["commitments"]
                details = ExecutionReceiptDetails(
                    "execution", metadata["decisionReceiptId"], execution_id.hex(), profile["id"], int(profile["version"]),
                    decision_metadata["decision"], metadata["outcome"], tuple(decision_metadata["reasons"]), decided_at,
                    completed_at, decision_signer, _signer(metadata["executionSigner"]), common["proof"], common["action"],
                    common["context"], common["principalStatus"], common["grantStatus"], execution_commitments["executionLease"],
                    execution_commitments["command"], execution_commitments.get("result"),
                )
            return VerifiedReceipt(_VERIFIED_TOKEN, source, details)
        except (AssertionError, KeyError, TypeError, ValueError, UnicodeDecodeError, json.JSONDecodeError):
            return RejectedReceipt("rejected", error_info("core.receipt-malformed"))


def _verify_with_anchors(kind: str, attested: bytes, receipt_id: bytes, role: ReceiptSignerRole, trust: ReceiptTrustPolicy) -> Dict[str, Any]:
    attempted = False
    for anchor in trust._anchors:
        if anchor.role != role:
            continue
        attempted = True
        try:
            encoded = verify_pinned_receipt_v1(kind, attested, receipt_id, anchor.principal, anchor.verification_method, anchor.suite, anchor.public_key)
            value = json.loads(bytes(encoded).decode("utf-8"))
            if type(value) is dict:
                return value
        except Exception:
            continue
    if not attempted:
        raise ValueError("receipt signer is untrusted")
    raise ValueError("receipt signature is invalid")


_VERIFIER_TOKEN = object()


_VERIFIER = _Verifier(_VERIFIER_TOKEN)


def verify(value: VerificationInput, /) -> VerificationResult:
    return _VERIFIER.verify(value)


def verify_many(values: Iterable[VerificationInput], /) -> Tuple[VerificationResult, ...]:
    return _VERIFIER.verify_many(values)


def inspect(value: VerificationResult, /) -> VerificationInspection:
    return _VERIFIER.inspect(value)


def verify_receipt(
    receipt: Union[Receipt, bytes], /, *, trust: ReceiptTrustPolicy,
) -> ReceiptVerification:
    return _VERIFIER.verify_receipt(
        receipt=receipt,
        trust=trust,
    )


def _project(native: NativeVerificationResult, correlation_id: str) -> VerificationResult:
    metrics = VerificationMetrics(*native.metrics)
    common = (native.code, native.stage, correlation_id, metrics, native.required_configuration, bytes(native.local_configuration), bytes(native.result_cbor))
    if native.kind == "authorized":
        return AuthorizedVerification("authorized", *common)
    code = "core.authorization-denied" if native.kind == "denied" else "core.authorization-indeterminate"
    return UnsuccessfulVerification(native.kind, *common, error_info(code, correlation_id=correlation_id))


def _signer(value: Dict[str, Any]) -> ReceiptSignerInfo:
    return ReceiptSignerInfo(value["principal"], value["verificationMethod"], value["suite"])


__all__ = [
    "ApprovalInspection", "AuthorizedVerification", "DecisionReceiptDetails", "ExecutionReceiptDetails", "IndeterminateReceipt", "ReceiptEnvelopeDetails", "ReceiptProfile", "ReceiptSignerInfo", "ReceiptTrustAnchor", "ReceiptTrustPolicy", "ReceiptVerification", "RejectedReceipt", "UnsuccessfulVerification", "VerificationInput", "VerificationInspection", "VerificationMetrics", "VerificationResult", "VerifiedReceipt", "inspect", "pinned_receipt_trust", "verify", "verify_many", "verify_receipt",
]
